### PR TITLE
Port propertize

### PR DESCRIPTION
**Pull request by shaleh**

### CODE REVIEW

**Comment by shaleh** [comment]
Grumble. My branch was out of sync. So I merged in from main. It builds locally. What did I miss?

**Comment by shaleh** [comment]
The logs complaints about:

In toplevel form:

src/undo-tests.el:262:6:Warning: `delete-forward-char' is for interactive use

only; use `delete-char' instead.

src/undo-tests.el:317:6:Warning: `delete-forward-char' is for interactive use

only; use `delete-char' instead.

src/undo-tests.el:353:47:Warning: `delete-forward-char' is for interactive use

only; use `delete-char' instead.

src/undo-tests.el:370:13:Warning: `delete-forward-char' is for interactive use

only; use `delete-char' instead.

src/undo-tests.el:394:19:Warning: `delete-forward-char' is for interactive use

only; use `delete-char' instead.

src/undo-tests.el:433:11:Warning: `delete-forward-char' is for interactive use

only; use `delete-char' instead.

GEN src/undo-tests.log

ELC src/xml-tests.elc

GEN src/xml-tests.log

make[3]: Leaving directory `/home/travis/build/Wilfred/remacs/test'

make[2]: [check-doit] Error 2 (ignored)

SUMMARY OF TEST RESULTS

-----------------------

Files examined: 175

Ran 2506 tests, 2438 results as expected, 23 unexpected, 45 skipped

6 files contained unexpected results:

lisp/replace-tests.log

lisp/register-tests.log

lisp/help-fns-tests.log

lisp/emacs-lisp/tabulated-list-test.log

lisp/emacs-lisp/package-tests.log

lisp/buff-menu-tests.log

make[2]: *** [check-doit] Error 1

make[2]: Leaving directory `/home/travis/build/Wilfred/remacs/test'

make[1]: *** [check] Error 2

make[1]: Leaving directory `/home/travis/build/Wilfred/remacs/test'

make: *** [check] Error 2

**Comment by DavidDeSimone** [comment]
@shaleh Thank you for this PR. So this PR is failing travis because the elisp tests are failing. You can run those tests locally by running `make check` in the same directory you run `make`. If a test fails, it will record the stack trace of the failure in a file named <test name>.log. 

For example, there is a failure in a test case in register-tests.el. From the output of travis, we have this stack trace:
```
make[3]: *** [lisp/register-tests.log] Error 1
 ELC lisp/replace-tests.elc
 GEN lisp/replace-tests.log
Running 16 tests (2017-10-22 15:59:53+0000)
Test occur-test-00 backtrace:
 signal(error ("Attempt to modify read-only object" "\n"))
 apply(signal (error ("Attempt to modify read-only object" "\n")))
 (setq value-12 (apply fn-10 args-11))
 (unwind-protect (setq value-12 (apply fn-10 args-11)) (setq form-des
 (if (unwind-protect (setq value-12 (apply fn-10 args-11)) (setq form
 (let (form-description-14) (if (unwind-protect (setq value-12 (apply
 (let ((value-12 'ert-form-evaluation-aborted-13)) (let (form-descrip
 (let* ((fn-10 (function equal)) (args-11 (condition-case err (let ((
 (let (replace-occur-hook) (let* ((fn-10 (function equal)) (args-11 (
 (lambda nil (let (replace-occur-hook) (let* ((fn-10 (function equal)
 ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
 ert-run-test(#s(ert-test :name occur-test-00 :documentation "Test el
 ert-run-or-rerun-test(#s(ert--stats :selector (not (tag :expensive-t
 ert-run-tests((not (tag :expensive-test)) #f(compiled-function (even
 ert-run-tests-batch((not (tag :expensive-test)))
 ert-run-tests-batch-and-exit((not (tag :expensive-test)))
 eval((ert-run-tests-batch-and-exit '(not (tag :expensive-test))))
 command-line-1(("-L" ":." "-l" "ert" "-l" "lisp/replace-tests.el" "-
 command-line()
 normal-top-level()
Test occur-test-00 condition:
 (error "Attempt to modify read-only object" "
")
```  

This reads that lisp/replace-tests.elc has a test case `occur-test-00` that failed with an error "Attempt to modify read-only object". You can find this stack in the travis output, if you search for 'replace-tests', and in test/lisp/replace-tests.log if you run the tests locally.

File: rust_src/src/editfns.rs [hunk 1]
```diff
@@ -182,3 +182,43 @@ pub fn char_after(mut pos: LispObject) -> LispObject {
...
+    }
+
+    let mut it = args.iter();
+
+    let first = it.next().unwrap(); // safe, there is at least 1 arg
+    if !first.is_string() {
+        wrong_type!(Qstringp, *first);
+    }
+
+    let copy = first.clone();
```
Comment (DavidDeSimone) [inline rust_src/src/editfns.rs]: Unfortunately, this is not doing what you would expect. You are calling the 'clone' method of `LispObject`. `LispObject`'s are just tagged pointers, so this is really just a pointer copy. You will want to copy the C layer's call to Fcopy_sequence.

File: rust_src/src/editfns.rs [hunk 2]
```diff
@@ -182,3 +182,43 @@ pub fn char_after(mut pos: LispObject) -> LispObject {
...
+pub fn propertize(args: &mut [LispObject]) -> LispObject {
+    /* Number of args must be odd. */
+    if args.len() & 1 == 0 {
+        error!("Wrong number of arguments");
+    }
+
+    let mut it = args.iter();
+
+    let first = it.next().unwrap(); // safe, there is at least 1 arg
+    if !first.is_string() {
```
Comment (DavidDeSimone) [inline rust_src/src/editfns.rs]: It should be fine to just do a `first.as_string_or_error()` here, and not use the result.

File: rust_src/src/editfns.rs [hunk 3]
```diff
@@ -182,3 +182,43 @@ pub fn char_after(mut pos: LispObject) -> LispObject {
...
+        wrong_type!(Qstringp, *first);
+    }
+
+    let copy = first.clone();
+    let mut properties = Qnil;
```
Comment (DavidDeSimone) [inline rust_src/src/editfns.rs]: We prefer to use LispObject::constant_nil() over a direct usage of Qnil.

File: rust_src/src/editfns.rs (hunk 1 above)
Comment (shaleh) [reply rust_src/src/editfns.rs]: Ah, I had used Fcopy_sequence and then thought I was doing the right thing by using the Rust method. Easy fix. Thanks.

File: rust_src/src/editfns.rs (hunk 2 above)
Comment (shaleh) [reply rust_src/src/editfns.rs]: Ok, I was trying to keep the C code's error.

File: rust_src/src/editfns.rs (hunk 3 above)
Comment (shaleh) [reply rust_src/src/editfns.rs]: I did not because properties is always a C Lisp_Object, not a Rust LispObject. This leads to type errors.

Comment (shaleh) [reply rust_src/src/editfns.rs]: I can add a comment so people see it. Eventually this will go away once Fcons is ported.

**Comment by shaleh** [comment]
Thanks for the pointer to `make check`. I ran it before switched from Fcopy_sequence and forgot to run it again. My mistake. I am fixing the PR currently.

File: rust_src/src/editfns.rs (hunk 2 above)
Comment (shaleh) [reply rust_src/src/editfns.rs]: If I use `as_string_or_error` I can no long use the `as_raw` method in the call to Fcopy_sequence.

Comment (DavidDeSimone) [reply rust_src/src/editfns.rs]: Good point, I think it's fine to keep it the way it was then.

Comment (shaleh) [reply rust_src/src/editfns.rs]: I think the new version satisfies the expected style.

File: rust_src/src/editfns.rs [hunk 4]
```diff
@@ -182,3 +182,44 @@ pub fn char_after(mut pos: LispObject) -> LispObject {
         }
     }
 }
+
+/// Return a copy of STRING with text properties added.
+/// First argument is the string to copy.
+/// Remaining arguments form a sequence of PROPERTY VALUE pairs for text
+/// properties to add to the result.
+/// usage: (propertize STRING &rest PROPERTIES)  */
```
Comment (Wilfred) [inline rust_src/src/editfns.rs]: Trailing C comment here.

Comment (shaleh) [reply rust_src/src/editfns.rs]: Ooops :-)

I can fix that real quick.

**Comment by Wilfred** [comment]
Looks good!

**Comment by Wilfred** [comment]
@shaleh I've added you as a collaborator, so you have the power to review and approve PRs if you wish! :bowtie:

**Comment by shaleh** [comment]
Thanks!

**Pull request merged**